artemis-server/src/main/java/org/apache/activemq/artemis/core/server/metrics/MetricsManager.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.activemq.artemis.core.server.metrics;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.ToDoubleFunction;
+
+import io.micrometer.core.instrument.Gauge;
+import io.micrometer.core.instrument.Meter;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Metrics;
+import io.micrometer.core.instrument.binder.jvm.JvmMemoryMetrics;
+import org.apache.activemq.artemis.api.core.management.ResourceNames;
+import org.apache.activemq.artemis.core.server.ActiveMQServerLogger;
+
+public class MetricsManager {
```

Comment:
That's a huge comment; API change proposal...
```java
   @FunctionalInterface
   public interface MetricGaugeBuilder {
      void register(String metricName, Object state, ToDoubleFunction f, String description);
   }

   public void registerQueueGauge(String address, String queue, Consumer<MetricGaugeBuilder> builder) {
      final List<Meter> newMeters = new ArrayList<>();
      builder.accept((metricName, state, f, description) -> {
         Meter meter = Gauge
            .builder("artemis." + metricName, state, f)
            .tag("broker", brokerName)
            .tag("address", address)
            .tag("queue", queue)
            .description(description)
            .register(meterRegistry);
         newMeters.add(meter);
      });
      final String resource = ResourceNames.QUEUE + queue;
      this.meters.compute(resource, (s, meters) -> {
         if (meters == null) {
            meters = new ArrayList<>();
         }
         meters.addAll(newMeters);
         return meters;
      });
   }

   public void registerAddressGauge(String address, Consumer<MetricGaugeBuilder> builder) {
      final List<Meter> newMeters = new ArrayList<>();
      builder.accept((metricName, state, f, description) -> {
         Meter meter = Gauge
            .builder("artemis." + metricName, state, f)
            .tag("broker", brokerName)
            .tag("address", address)
            .description(description)
            .register(meterRegistry);
         newMeters.add(meter);
      });
      final String resource = ResourceNames.ADDRESS + address;
      this.meters.compute(resource, (s, meters) -> {
         if (meters == null) {
            meters = new ArrayList<>();
         }
         meters.addAll(newMeters);
         return meters;
      });
   }

   public void registerBrokerGauge(Consumer<MetricGaugeBuilder> builder) {
      final List<Meter> newMeters = new ArrayList<>();
      builder.accept((metricName, state, f, description) -> {
         Meter meter = Gauge
            .builder("artemis." + metricName, state, f)
            .tag("broker", brokerName)
            .description(description)
            .register(meterRegistry);
         newMeters.add(meter);
      });
      final String resource = ResourceNames.BROKER + "." + brokerName;
      this.meters.compute(resource, (s, meters) -> {
         if (meters == null) {
            meters = new ArrayList<>();
         }
         meters.addAll(newMeters);
         return meters;
      });
   }
```

That could be used with:
```java
if (server != null && server.getMetricsManager() != null) {
         server.getMetricsManager().registerQueueGauge(queueName, addressName, builder -> {
            builder.register("message.count", pendingMetrics, metrics -> Double.valueOf(pendingMetrics.getMessageCount()), QueueControl.MESSAGE_COUNT_DESCRIPTION);
            builder.register(//...other useful metrics
```

Reply:
Maybe it's just me, but I think the existing implementation is easier to read.  :)

Is there a particular benefit to your implementation? I'm assuming it has better performance for some esoteric reason. :)

Reply:
Maybe :P It has a better behaviour under load, because it won't call many times put on the concurrent hash map, just once...

Reply:
And it is more safe concurrently, given that the overall configuration is consider an atomic operation